utils/near-framed-read/src/framed_impl.rs

```diff
@@ -0,0 +1,144 @@
+use std::borrow::BorrowMut;
+use std::future::Future;
+use std::io::{self};
+use std::mem::MaybeUninit;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+use bytes::{BufMut, BytesMut};
+use futures_core::{ready, Stream};
+use futures_timer::Delay;
+use log::trace;
+use pin_project_lite::pin_project;
+use tokio::io::{AsyncRead, ReadBuf};
+
+use crate::decoder::Decoder;
+
+pin_project! {
+    #[derive(Debug)]
+    pub(crate) struct FramedImpl<T, U, State> {
+        #[pin]
+        pub(crate) inner: T,
+        pub(crate) state: State,
+        pub(crate) codec: U,
+        pub(crate) delay: Delay,
+        pub(crate) duration: Duration,
+        pub(crate) bytes: i64,
+        pub(crate) bytes_left: i64,
+    }
+}
+
+const INITIAL_CAPACITY: usize = 8 * 1024;
+
+pub(crate) struct ReadFrame {
+    pub(crate) eof: bool,
+    pub(crate) is_readable: bool,
+    pub(crate) buffer: BytesMut,
+}
+
+impl Default for ReadFrame {
+    fn default() -> Self {
+        Self { eof: false, is_readable: false, buffer: BytesMut::with_capacity(INITIAL_CAPACITY) }
+    }
+}
+
+impl<T, U, R> Stream for FramedImpl<T, U, R>
+where
+    T: AsyncRead,
+    U: Decoder,
+    R: BorrowMut<ReadFrame>,
+{
+    type Item = Result<U::Item, U::Error>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut pinned = self.project();
+        let state: &mut ReadFrame = pinned.state.borrow_mut();
+        loop {
+            // Check if we should apply throttling
+            if *pinned.bytes > 0 && *pinned.bytes_left < 0 {
+                match unsafe { Pin::new_unchecked(&mut pinned.delay) }.poll(cx) {
+                    Poll::Ready(_) => {
+                        *pinned.bytes_left += *pinned.bytes;
+                        pinned.delay.reset(*pinned.duration);
+                    }
+
+                    Poll::Pending => {
+                        return Poll::Pending;
+                    }
+                }
+                // We may wait more than 1 duration, to make sure we don't go over read limit.
+                if *pinned.bytes_left < 0 {
+                    return Poll::Pending;
```

Comment:
Hm, this feels like a bug -- we return a `Pending` here, but we don't seem to arrange for `waker` to be called. So if we hit this code path, the executor might never poll this future. Not entirely sure what's the proper fix is, but maybe just `continue;` instead of `return`?